### ISSUE TITLE
Support H2 v1.4.198 array syntax

### ISSUE THREAD

**Issue by ivangreene** [jOOQ/jOOQ#8434]
### Expected behavior and actual behavior:
Selecting from an array as a table should work in H2 versions 1.4.198 and 1.4.199, but fails due to the syntax changing from `(1,2)` to `array[1,2]`. This causes H2 to throw an exception like the following:
```
org.jooq.exception.DataAccessException: SQL [select "v"."c1" from ((select null "c1" where 1 = 0) union all (select * from (values ((cast(? as int), cast(? as int), cast(? as int), cast(? as int))), ((cast(? as int), cast(? as int)))) "v")) "v"]; Unknown data type: "ROW"; SQL statement:
SELECT
    "TABLE".C1
FROM TABLE(C1 ROW=ROW (ROW (CAST(?1 AS INTEGER), CAST(?2 AS INTEGER), CAST(?3 AS INTEGER), CAST(?4 AS INTEGER)), ROW (CAST(?5 AS INTEGER), CAST(?6 AS INTEGER)))) [50004-198]
```
### Steps to reproduce the problem (if possible, create an MCVE: https://github.com/jOOQ/jOOQ-mcve):
MCVE available here: https://github.com/ivangreene/jOOQ-mcve/blob/f82fa659fd471896b67ee9619ca8ba23f798bd79/src/test/java/org/jooq/mcve/test/MCVETest.java
### Versions:

- jOOQ: 3.11.7
- Java: 8
- Database (include vendor): H2 v1.4.198 and v1.4.199
- OS: MacOS 10.14.2
- JDBC Driver (include name if inofficial driver):

**Comment by ivangreene** [comment]
A SQL example:

This will work up to and including H2 v1.4.997, but not 1.4.998 or 1.4.999:
```sql
select "v"."c1" from
((select null "c1" where 1 = 0)
  union all
  (select * from (values ((1, 2)), ((3, 4))) "v")) "v"
where 3 in (table(i int = "v"."c1"));
```

This syntax works in 1.4.998 and 1.4.999 only:
```sql
select "v"."c1" from
((select null "c1" where 1 = 0)
  union all
  (select * from (values (array[1, 2]), (array[3, 4])) "v")) "v"
where 3 in (table(i int = "v"."c1"));
```

**Comment by katzyn** [comment]
@lukaseder
I think you need to generate standard array literals for H2 >= 1.4.198 and legacy H2-specific literals for all older versions.

1.4.198 and 1.4.199 still support legacy literals is some common places, but not here and such support may be removed in newer versions due to some minor issues with it. Perhaps table value constructor in H2 should also forbid usage of row values inside its top-level row values with some sane error message, such construction is not valid.

**Comment by lukaseder** [comment]
Thank you all for your comments. Indeed, I've noticed quite a few things breaking when upgrading to 1.4.198 / 1.4.199 in this area. Indeed, jOOQ used to emulate row value expressions using array literals, e.g. to do predicates like this:

```sql
-- Comparing arrays in H2
(a, b) IN (SELECT (x, y) FROM t)

-- ... instead of comparing row value expressions
(a, b) IN (SELECT x, y FROM t)
```

I haven't noticed too many regressions in the array handling code yet, but that may also be due to the fact that those aren't very extensively integration tested, given that arrays are not SQL's most popular feature.

Given that 1.4.198 is quite a major release, I do think it makes sense to distinguish between version numbers in jOOQ. Or, alternatively, break compatibility as well and make jOOQ 3.11 support 1.4.197 and less and jOOQ 3.12 support 1.4.198 and more. I have not decided yet, what's better in the case of H2.

I still wish H2 would implement semver, though :) So many new features and breaking changes in what appears to be a random patch release is surprising...

I will review the various issues that are related to the `(a, b)` syntax (previously arrays now row value expressions) and get in touch on H2's issue tracker if there are any doubts.

**Comment by katzyn** [comment]
1.4.199 is a patch release, but 1.4.198 is a big release with a lot of changes. Version numbers are just numbers here, they don't have hidden sense. Unfortunately, we don't have free time to support old releases in any way.

**Comment by lukaseder** [comment]
Given that H2 currently doesn't support semver, nor old versions, we won't support old versions of H2 either. jOOQ 3.12 will officially support 1.4.199, whereas jOOQ 3.11 will support 1.4.197

**Comment by lukaseder** [comment]
I have implemented also #8502 to re-support row value expression predicates in jOOQ 3.12

**Comment by lukaseder** [comment]
Implemented in jOOQ 3.12, which will generate array literals as `array[a, b]` instead of `(a, b)`. Because this is an incompatible change (unsupporting H2 1.4.197 syntax elements), it will not be backported to jOOQ 3.11.

Thanks for all your help, everyone!
